hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/MapIndexScanP.java

```diff
@@ -0,0 +1,464 @@
+        }
+    }
+
+    public static final class MapIndexScanProcessorMetaSupplier implements ProcessorMetaSupplier, DataSerializable {
+
+        private MapIndexScanMetadata indexScanMetadata;
+
+        @SuppressWarnings("unused")
+        private MapIndexScanProcessorMetaSupplier() {
+        }
+
+        public MapIndexScanProcessorMetaSupplier(
+                @Nonnull MapIndexScanMetadata indexScanMetadata
+        ) {
+            this.indexScanMetadata = indexScanMetadata;
+        }
+
+        @Override
+        @Nonnull
+        public Function<Address, ProcessorSupplier> get(@Nonnull List<Address> addresses) {
+            return address -> new MapIndexScanProcessorSupplier(indexScanMetadata);
+        }
+
+        @Override
+        public void writeData(ObjectDataOutput out) throws IOException {
+            out.writeObject(indexScanMetadata);
+        }
+
+        @Override
+        public void readData(ObjectDataInput in) throws IOException {
+            indexScanMetadata = in.readObject();
+        }
+    }
+
+    public static final class MapIndexScanProcessorSupplier implements ProcessorSupplier, DataSerializable {
+
+        private MapIndexScanMetadata indexScanMetadata;
+
+        @SuppressWarnings("unused")
+        private MapIndexScanProcessorSupplier() {
+        }
+
+        private MapIndexScanProcessorSupplier(@Nonnull MapIndexScanMetadata indexScanMetadata) {
+            this.indexScanMetadata = indexScanMetadata;
+        }
+
+        @Override
+        @Nonnull
+        public List<Processor> get(int count) {
+            return IntStream.range(0, count)
+                    .mapToObj(i -> new MapIndexScanP(indexScanMetadata))
+                    .collect(toList());
+        }
+
+        @Override
+        public void writeData(ObjectDataOutput out) throws IOException {
+            out.writeObject(indexScanMetadata);
+        }
+
+        @Override
+        public void readData(ObjectDataInput in) throws IOException {
+            indexScanMetadata = in.readObject();
+        }
+    }
+
+    static final class LocalMapIndexReader extends AbstractIndexReader<MapFetchIndexOperationResult, QueryableEntry<?, ?>> {
+
+        private HazelcastInstance hazelcastInstance;
+        private String indexName;
+
+        private LocalMapIndexReader(@Nonnull HazelcastInstance hzInstance,
+                                    @Nonnull InternalSerializationService serializationService,
+                                    @Nonnull MapIndexScanMetadata indexScanMetadata) {
+            super(indexScanMetadata.getMapName(), MapFetchIndexOperationResult::getEntries);
+            this.hazelcastInstance = hzInstance;
+            this.indexName = indexScanMetadata.getIndexName();
+            this.serializationService = serializationService;
+        }
+
+        public static LocalMapIndexReader create(@Nonnull HazelcastInstance hzInstance,
```

Comment:
`AbstractIndexReader` has only one subclass, `LocalMapIndexReader`. Moreover, `LocalMapIndexReader` is `DataSerializable` but does not seem to be ever serialized, it does not even have default constructor.

Reply:
> Moreover, `LocalMapIndexReader` is `DataSerializable` but does not seem to be ever serialized, it does not even have default constructor.

Agree, removed.

> `AbstractIndexReader` has only one subclass, `LocalMapIndexReader`

@viliam-durina, @hakanmemisoglu  and me had the possibility of reading indices from remote cluster. To be honest, I don't remember final decision, but I left `AbstractIndexReader` as an abstract class for reading indices from remote cluster.